File: rabbitmq/cpp/sender_async_test.cpp
#include <rabbitmq-c/amqp.h>
#include <rabbitmq-c/tcp_socket.h>
#include <iostream>
#include <fstream>
#include <string>
#include <future>
#include <vector>
#include "../../utils/cpp/stats_collector.hpp"
#include "../../utils/cpp/test_data_loader.hpp"
#include "../../utils/cpp/message_helpers.hpp"

using json = nlohmann::json;
using messaging::MessageEnvelope;
using message_helpers::get_current_time_ms;

struct TaskResult {
    bool success;
    std::string message_id;
    long long duration;
    std::string error;
};

TaskResult send_message_task(const json& item) {
    TaskResult res;
    res.success = false;
    res.message_id = message_helpers::extract_message_id(item);
    res.duration = 0;

    amqp_connection_state_t conn = amqp_new_connection();
    amqp_socket_t *socket = amqp_tcp_socket_new(conn);
    
    if (amqp_socket_open(socket, "localhost", 5672) != 0) {
        res.error = "Connection failed";
        amqp_destroy_connection(conn);
        return res;
    }

    amqp_login(conn, "/", 0, 131072, 0, AMQP_SASL_METHOD_PLAIN, "guest", "guest");
    amqp_channel_open(conn, 1);
    
    // Subscribe to direct reply queue
    amqp_basic_consume(conn, 1, amqp_cstring_bytes("amq.rabbitmq.reply-to"), amqp_empty_bytes, 0, 1, 0, amqp_empty_table);

    int target = item.value("target", 0);
    std::string queue_name = "test_queue_" + std::to_string(target);
    std::string reply_queue = "amq.rabbitmq.reply-to";

    long long msg_start = get_current_time_ms();

    // Create and send message
    MessageEnvelope envelope = message_helpers::create_data_envelope(item);
    std::string body = message_helpers::serialize_envelope(envelope);

    amqp_basic_properties_t props;
    props._flags = AMQP_BASIC_CONTENT_TYPE_FLAG | AMQP_BASIC_REPLY_TO_FLAG | AMQP_BASIC_CORRELATION_ID_FLAG;
    props.content_type = amqp_cstring_bytes("application/octet-stream");
    props.reply_to = amqp_cstring_bytes(reply_queue.c_str());
    props.correlation_id = amqp_cstring_bytes(res.message_id.c_str());

    amqp_basic_publish(conn, 1, amqp_empty_bytes, amqp_cstring_bytes(queue_name.c_str()),
                      0, 0, &props, amqp_cstring_bytes(body.c_str()));

    // Wait for reply
    struct timeval timeout = {0, 100000};  // 100ms
    amqp_envelope_t reply_envelope;
    amqp_rpc_reply_t rpc_res = amqp_consume_message(conn, &reply_envelope, &timeout, 0);

    if (rpc_res.reply_type == AMQP_RESPONSE_NORMAL) {
        std::string reply_str((char*)reply_envelope.message.body.bytes, reply_envelope.message.body.len);
        
        MessageEnvelope resp_envelope;
        if (message_helpers::parse_envelope(reply_str, resp_envelope) && 
            message_helpers::is_valid_ack(resp_envelope, res.message_id)) {
            res.duration = get_current_time_ms() - msg_start;
            res.success = true;
        } else {
            res.error = "Invalid ACK";
        }
        amqp_destroy_envelope(&reply_envelope);
    } else {
        res.error = "Timeout";
    }

    amqp_channel_close(conn, 1, AMQP_REPLY_SUCCESS);
    amqp_connection_close(conn, AMQP_REPLY_SUCCESS);
    amqp_destroy_connection(conn);

    return res;
}

int main() {
    auto test_data = test_data_loader::loadTestData();

    MessageStats stats;
    stats.set_metadata({
        {"service", "RabbitMQ"},
        {"language", "C++"},
        {"async", true}
    });
    long long start_time = get_current_time_ms();

    std::cout << " [x] Starting ASYNC transfer of " << test_data.size() << " messages..." << std::endl;

    std::vector<std::future<TaskResult>> futures;
    for (auto& item : test_data) {
        futures.push_back(std::async(std::launch::async, send_message_task, item));
    }

    for (auto& fut : futures) {
        TaskResult res = fut.get();
        if (res.success) {
            stats.record_message(true, res.duration);
            std::cout << " [OK] Message " << res.message_id << " acknowledged" << std::endl;
        } else {
            stats.record_message(false);
            std::cout << " [FAILED] Message " << res.message_id << ": " << res.error << std::endl;
        }
    }

    long long end_time = get_current_time_ms();
    stats.set_duration(start_time, end_time);
    
    json report = stats.get_stats();

    std::cout << "\nTest Results (ASYNC):" << std::endl;
    std::cout << "total_sent: " << stats.sent_count << std::endl;
    std::cout << "total_received: " << stats.received_count << std::endl;
    std::cout << "duration_ms: " << stats.get_duration_ms() << std::endl;

    std::ofstream rf("logs/report.txt", std::ios::app);
    if (rf.good()) {
        rf << report.dump() << std::endl;
        rf.close();
    }

    return 0;
}

File: rabbitmq/python/sender_async_test.py
#!/usr/bin/env python3
"""RabbitMQ Python Sender - Async"""
import sys
import json
import asyncio
import aio_pika
from pathlib import Path

# Add utils to path
repo_root = Path(__file__).parent.parent.parent
sys.path.insert(0, str(repo_root / 'utils' / 'python'))

from message_helpers import *
from test_data_loader import load_test_data
from stats_collector import MessageStats


async def send_message_task(channel, item):
    """Send a single message asynchronously."""
    result = {'success': False, 'message_id': '', 'duration': 0, 'error': ''}
    
    try:
        message_id = extract_message_id(item)
        result['message_id'] = message_id
        target = item.get('target', 0)
        
        queue_name = f"test_queue_{target}"
        msg_start = get_current_time_ms()
        
        # Declare reply queue (exclusive)
        reply_queue = await channel.declare_queue(exclusive=True)
        
        # Create and send message
        envelope = create_data_envelope(item)
        body = serialize_envelope(envelope)
        
        future = channel.default_exchange.publish(
            aio_pika.Message(
                body=body,
                content_type='application/octet-stream',
                correlation_id=message_id,
                reply_to=reply_queue.name
            ),
            routing_key=queue_name
        )
        await future
        
        # Wait for reply
        try:
            async with reply_queue.iterator() as queue_iter:
                async for message in queue_iter:
                    async with message.process():
                        resp_envelope = parse_envelope(message.body)
                        if is_valid_ack(resp_envelope, message_id):
                            result['duration'] = get_current_time_ms() - msg_start
                            result['success'] = True
                        else:
                            result['error'] = 'Invalid ACK'
                        break # Only expect one response
        except asyncio.TimeoutError:
            result['error'] = 'Timeout'
            
    except Exception as e:
        result['error'] = str(e)
    
    return result


async def run():
    test_data = load_test_data()
    
    stats = MessageStats()
    stats.set_metadata({
        'service': 'RabbitMQ',
        'language': 'Python',
        'async': True
    })
    start_time = get_current_time_ms()
    
    print(f" [x] Starting ASYNC transfer of {len(test_data)} messages...")
    
    connection = await aio_pika.connect_robust("amqp://guest:guest@localhost/")
    
    async with connection:
        # Create channel
        channel = await connection.channel()
        
        # We need to process messages one by one or in batches as asyncio.gather 
        # with dynamic reply queues can be resource intensive.
        # For simplicity in this test pattern, we'll confirm strictly sequential async dispatch/wait
        # To make it truly parallel, we'd need shared reply queues/correlation ID map.
        # But mirroring C++ async test structure which uses futures:
        
        tasks = [send_message_task(await connection.channel(), item) for item in test_data]
        results = await asyncio.gather(*tasks)
        
        for result in results:
            if result['success']:
                stats.record_message(True, result['duration'])
                print(f" [OK] Message {result['message_id']} acknowledged")
            else:
                stats.record_message(False)
                print(f" [FAILED] Message {result['message_id']}: {result['error']}")
    
    end_time = get_current_time_ms()
    stats.set_duration(start_time, end_time)
    
    report = stats.get_stats()
    
    print("\nTest Results (ASYNC):")
    print(f"total_sent: {stats.sent_count}")
    print(f"total_received: {stats.received_count}")
    print(f"duration_ms: {stats.get_duration_ms()}")
    
    with open('logs/report.txt', 'a') as f:
        f.write(json.dumps(report) + '\n')


def main():
    asyncio.run(run())


if __name__ == "__main__":
    main()
